src/hypercube.c

```diff
@@ -0,0 +1,181 @@
+ * Sort the hypercubes slices in ascending dimension ID order. This allows us to
+ * iterate slices in a consistent order.
+ */
+void
+hypercube_slice_sort(Hypercube *hc)
+{
+	qsort(hc->slices, hc->num_slices, sizeof(DimensionSlice *), cmp_slices_by_dimension_id);
+}
+
+DimensionSlice *
+hypercube_get_slice_by_dimension_id(Hypercube *hc, int32 dimension_id)
+{
+	DimensionSlice slice = {
+		.fd.dimension_id = dimension_id,
+	};
+	void	   *ptr = &slice;
+
+	ptr = bsearch(&ptr, hc->slices, hc->num_slices,
+				  sizeof(DimensionSlice *), cmp_slices_by_dimension_id);
+
+	if (NULL == ptr)
+		return NULL;
+
+	return *((DimensionSlice **) ptr);
+}
+
+/*
+ * Given a set of constraints, build the corresponding hypercube.
+ */
+Hypercube *
+hypercube_from_constraints(ChunkConstraint constraints[], int16 num_constraints)
+{
+	Hypercube  *hc = hypercube_alloc(num_constraints);
+	int			i;
+
+	for (i = 0; i < num_constraints; i++)
+	{
+		DimensionSlice *slice = dimension_slice_scan_by_id(constraints[i].fd.dimension_slice_id);
+
+		Assert(slice != NULL);
+		hc->slices[hc->num_slices++] = slice;
+	}
+
+	hypercube_slice_sort(hc);
+	return hc;
+}
+
+/*
+ * Calculate the hypercube that encloses the given point.
+ *
+ * The hypercube's configuration depends on the current partitioning in each
+ * dimension of the N-dimensional hyperspace, including any alignment
+ * requirements.
+ */
+Hypercube *
+hypercube_calculate_from_point(Hyperspace *hs, Point *p)
+{
+	Hypercube  *cube;
+	int			i;
+
+	cube = hypercube_alloc(hs->num_dimensions);
+
+	/* For each dimension, calculate the hypercube's slice in that dimension */
+	for (i = 0; i < hs->num_dimensions; i++)
+	{
+		Dimension  *dim = &hs->dimensions[i];
+		int64		value = p->coordinates[i];
+		bool		found = false;
+
+		/*
+		 * If this is an aligned dimension, we'd like to reuse any existing
+		 * slice that covers the coordnate in the dimenion
+		 */
+		if (dim->fd.aligned)
+		{
+			DimensionVec *vec;
+
+			vec = dimension_slice_scan_limit(dim->fd.id, value, 1);
+
+			if (vec->num_slices > 0)
+			{
+				cube->slices[i] = vec->slices[0];
+				found = true;
+			}
+		}
+
+		if (!found)
+		{
+			/*
+			 * No existing slice found, or we are not aligning, so calculate
+			 * the range of a new slice
+			 */
+			cube->slices[i] = dimension_calculate_default_slice(dim, value);
+
+			/*
+			 * Check if there's already an existing slice with the calculated
+			 * range. If a slice already exists, use that slice's ID instead
+			 * of a new one.
+			 */
+			dimension_slice_scan_for_existing(cube->slices[i]);
+		}
+	}
+
+	cube->num_slices = hs->num_dimensions;
+
+	return cube;
+}
+
+/*
+ * Check if two hypercubes collide (overlap).
+ *
+ * This is basically an axis-aligned bounding box collision detection,
+ * generalized to N dimensions. We check for dimension slice collisions in each
+ * dimension and only if all dimensions collide there is a hypercube collision.
+ */
+bool
+hypercubes_collide(Hypercube *cube1, Hypercube *cube2)
```

Comment:
Can we Assert the same number of dimensions for both hypercubes?

Reply:
done.